docs/msi_with_credential_design.md

```diff
@@ -0,0 +1,233 @@
+- This credential is only valid for a short duration (1 hour) and must be used **immediately** to acquire an access token from ESTS.
+- This mechanism improves security by reducing the lifetime of sensitive authentication materials.
+
+## Steps for MSI V2 Authentication
+
+This section outlines the necessary steps to acquire an access token using the MSI V2 `/credential` endpoint. 
+
+### 1. Check for an Existing (Platform) Certificate
+- Search for a specific certificate (`devicecert.mtlsauth.local`) in `Cert:\LocalMachine\My`.
+- If found, extract its thumbprint and use it for authentication.
+
+### 2. Generate a New Certificate (if specific certificate is not found)
+- Create a new self-signed certificate with a 90-day validity.
+- Ensure the certificate has:
+  - Subject name `CN=mtls-auth` (name not final).
+  - Exportable key policy.
+
+### 3. Extract Certificate Data
+- Convert the certificate to a Base64-encoded string (`x5c`).
+- Format the JSON payload containing the certificate details for request authentication.
+
+### 4. Request MSI Credential
+- Send a POST request to the IMDS `/credential` endpoint with the certificate details.
+- The request must include:
+  - `Metadata: true` header.
+  - `X-ms-Client-Request-id` header with a GUID.
+  - JSON body containing the certificate's public key in `jwk` format.
+- Parse the response to extract:
+  - `regional_token_url`
+  - `tenant_id`
+  - `client_id`
+  - `credential` (short-lived credential).
+
+### 5. Request Access Token from ESTS
+- Construct the OAuth2 request body, including:
+  - `grant_type=client_credentials`
+  - `scope=https://management.azure.com/.default`
+  - `client_id` from the MSI response.
+  - `client_assertion` containing the short-lived credential.
+  - `client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer`.
+- Send a POST request to the `regional_token_url` with the certificate for mutual TLS (mTLS) authentication.
+
+### 6. Retrieve and Use Access Token
+- Parse the response to extract the `access_token`.
+- Use the access token to authenticate requests to Azure services.
+- Handle any errors that may occur during the token request.
+
+## End-to-End Script
+
+```powershell
+# Define certificate details
+$certSubject = "CN=mtls-auth"
+$certThumbprint = ""
+
+# Check for an existing valid certificate in LocalMachine\My
+$existingCert = Get-ChildItem -Path "Cert:\LocalMachine\My" | Where-Object { $_.Subject -like "*$certSubject*" -and $_.NotAfter -gt (Get-Date) }
+
+if ($existingCert) {
+    Write-Output "✅ Found existing valid certificate: $($existingCert.Subject)"
+    $cert = $existingCert
+} else {
+    Write-Output "❌ No valid certificate found. Creating a new self-signed certificate..."
+
+    # Create a new self-signed certificate
+    $cert = New-SelfSignedCertificate `
+        -Subject $certSubject `
+        -CertStoreLocation "Cert:\LocalMachine\My" `
```

Comment:
But in MSAL we shoudl store it in memory no?

Reply:
yes, added the code comment